File: src/year2024/day4/day4.rs
//! [Day 4: Ceres Search](https://adventofcode.com/2024/day/4)

use aoc::Coord;

type Grid = aoc::Grid<char>;

struct Puzzle {
    grid: Grid,
}

impl Puzzle {
    fn new(data: &str) -> Self {
        Self {
            grid: Grid::parse(data),
        }
    }

    /// Solve part one.
    fn part1(&self) -> u32 {
        let mut n = 0;
        let (sx, sy) = (self.grid.width(), self.grid.height());

        let mas = |i| match i {
            1 => 'M',
            2 => 'A',
            3 => 'S',
            _ => unreachable!(),
        };

        for (Coord { x, y }, _) in self.grid.iter().filter(|(_, p)| *p == &'X') {
            //
            if x <= sx - 4 && (1..4).all(|i| self.grid[(x + i, y)] == mas(i)) {
                n += 1;
            }
            if y <= sy - 4 && (1..4).all(|i| self.grid[(x, y + i)] == mas(i)) {
                n += 1;
            }

            if x >= 3 && (1..4).all(|i| self.grid[(x - i, y)] == mas(i)) {
                n += 1;
            }
            if y >= 3 && (1..4).all(|i| self.grid[(x, y - i)] == mas(i)) {
                n += 1;
            }

            if x <= sx - 4 && y <= sy - 4 && (1..4).all(|i| self.grid[(x + i, y + i)] == mas(i)) {
                n += 1;
            }
            if x >= 3 && y >= 3 && (1..4).all(|i| self.grid[(x - i, y - i)] == mas(i)) {
                n += 1;
            }

            if x <= sx - 4 && y >= 3 && (1..4).all(|i| self.grid[(x + i, y - i)] == mas(i)) {
                n += 1;
            }
            if x >= 3 && y <= sy - 4 && (1..4).all(|i| self.grid[(x - i, y + i)] == mas(i)) {
                n += 1;
            }
        }
        n
    }

    /// Solve part two.
    fn part2(&self) -> u32 {
        let mut n = 0;
        let (sx, sy) = (self.grid.width(), self.grid.height());
        for x in 1..(sx - 1) {
            for y in 1..(sy - 1) {
                if self.grid[(x, y)] == 'A' {
                    let ul = self.grid[(x - 1, y - 1)];
                    let ur = self.grid[(x + 1, y - 1)];
                    let bl = self.grid[(x - 1, y + 1)];
                    let br = self.grid[(x + 1, y + 1)];

                    if ((ul == 'M' && br == 'S') || (ul == 'S' && br == 'M'))
                        && ((ur == 'M' && bl == 'S') || (ur == 'S' && bl == 'M'))
                    {
                        n += 1;
                    }
                }
            }
        }
        n
    }
}

/// # Panics
/// over malformed input
#[must_use]
pub fn solve(data: &str) -> (u32, u32) {
    let puzzle = Puzzle::new(data);
    (puzzle.part1(), puzzle.part2())
}

pub fn main() {
    let args = aoc::parse_args();
    args.run(solve);
}

#[cfg(test)]
mod test {
    use super::*;

    const SAMPLE_1: &str = include_str!("sample_1.txt");
    const SAMPLE_2: &str = include_str!("sample_2.txt");
    const SAMPLE_3: &str = include_str!("sample_3.txt");
    const SAMPLE_4: &str = include_str!("sample_4.txt");
    const SAMPLE_5: &str = include_str!("sample_5.txt");

    #[test]
    fn test_p1_1() {
        let puzzle = Puzzle::new(SAMPLE_1);
        assert_eq!(puzzle.part1(), 4);
    }

    #[test]
    fn test_p1_2() {
        let puzzle = Puzzle::new(SAMPLE_2);
        assert_eq!(puzzle.part1(), 18);

        let puzzle = Puzzle::new(SAMPLE_3);
        assert_eq!(puzzle.part1(), 18);
    }

    #[test]
    fn test_p2_1() {
        let puzzle = Puzzle::new(SAMPLE_4);
        assert_eq!(puzzle.part2(), 1);
    }

    #[test]
    fn test_p2_2() {
        let puzzle = Puzzle::new(SAMPLE_2);
        assert_eq!(puzzle.part2(), 9);

        let puzzle = Puzzle::new(SAMPLE_5);
        assert_eq!(puzzle.part2(), 9);
    }
}

File: src/year2024/day4/day4.go
// [Day 4: Ceres Search](https://adventofcode.com/2024/day/4)

package main

import (
	"fmt"
	"log"
	"os"
	"strings"
	"time"
)

func solve(data []byte) (int, int) {

	lines := strings.Split(strings.TrimSpace(string(data)), "\n")
	sy := len(lines)
	sx := len(lines[0])

	letters := func(x, y int) byte {
		if 0 <= x && x < sx && 0 <= y && y < sy {
			return lines[y][x]
		}
		return '.'
	}

	// Part 1
	part1 := 0
	for y := 0; y < sy; y++ {
		for x := 0; x < sx; x++ {
			if letters(x, y) == 'X' {
				// Check 8 directions for "MAS"
				if string([]byte{letters(x, y+1), letters(x, y+2), letters(x, y+3)}) == "MAS" {
					part1++
				}
				if string([]byte{letters(x+1, y), letters(x+2, y), letters(x+3, y)}) == "MAS" {
					part1++
				}
				if string([]byte{letters(x+1, y+1), letters(x+2, y+2), letters(x+3, y+3)}) == "MAS" {
					part1++
				}
				if string([]byte{letters(x+1, y-1), letters(x+2, y-2), letters(x+3, y-3)}) == "MAS" {
					part1++
				}
				if string([]byte{letters(x, y-1), letters(x, y-2), letters(x, y-3)}) == "MAS" {
					part1++
				}
				if string([]byte{letters(x-1, y), letters(x-2, y), letters(x-3, y)}) == "MAS" {
					part1++
				}
				if string([]byte{letters(x-1, y+1), letters(x-2, y+2), letters(x-3, y+3)}) == "MAS" {
					part1++
				}
				if string([]byte{letters(x-1, y-1), letters(x-2, y-2), letters(x-3, y-3)}) == "MAS" {
					part1++
				}
			}
		}
	}

	// Part 2
	part2 := 0
	for y := 0; y < sy; y++ {
		for x := 0; x < sx; x++ {
			if letters(x, y) == 'A' {
				w := string([]byte{
					letters(x-1, y-1),
					letters(x+1, y+1),
					letters(x+1, y-1),
					letters(x-1, y+1),
				})
				if w == "MSMS" || w == "MSSM" || w == "SMMS" || w == "SMSM" {
					part2++
				}
			}
		}
	}

	return part1, part2
}

func main() {
	filename := "input.txt"
	elapsed := false

	for i := 1; i < len(os.Args); i++ {
		arg := os.Args[i]
		if arg == "--elapsed" {
			elapsed = true
		} else if !strings.HasPrefix(arg, "-") {
			filename = arg
		}
	}

	data, err := os.ReadFile(filename)
	if err != nil {
		log.Fatal(err)
	}

	if elapsed {
		start := time.Now()
		result1, result2 := solve(data)
		duration := time.Since(start)

		fmt.Println(result1)
		fmt.Println(result2)
		fmt.Printf("elapsed: %f ms\n", duration.Seconds()*1000.0)
	} else {
		result1, result2 := solve(data)

		fmt.Println(result1)
		fmt.Println(result2)
	}
}
